File: Algorithms/234_Palindrome_Linked_List/Palindrome_Linked_List.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
描述：回文链表
请判断一个链表是否为回文链表。

示例 1:

输入: 1->2
输出: false
示例 2:

输入: 1->2->2->1
输出: true
进阶：
你能否用 O(n) 时间复杂度和 O(1) 空间复杂度解决此题？

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/palindrome-linked-list
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
"""

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        """普通版, O(n)时间复杂度 O(n)空间复杂度"""
        result = []
        while head:
            result.append(head.val)
            head = head.next
        for i in range(len(result) // 2):
            if result[i] != result[len(result) - i - 1]:
                return False
        return True

    def isPalindrome2(self, head: ListNode) -> bool:
        """进阶版, O(n)时间复杂度 O(1)空间复杂度, 快慢双指针+反转链表"""
        # rev records the first half, need to set the same structure as fast, slow, hence later we have rev.next
        rev = None
        # initially slow and fast are the same, starting from head
        slow = fast = head
        while fast and fast.next:
            # fast traverses faster and moves to the end of the list if the length is odd
            fast = fast.next.next

            # take it as a tuple being assigned (rev, rev.next, slow) = (slow, rev, slow.next), hence the re-assignment of slow would not affect rev (rev = slow)
            rev, rev.next, slow = slow, rev, slow.next
        if fast:
            # fast is at the end, move slow one step further for comparison(cross middle one)
            slow = slow.next
        # compare the reversed first half with the second half
        while rev and rev.val == slow.val:
            slow = slow.next
            rev = rev.next

        # if equivalent then rev become None, return True; otherwise return False
        return not rev

File: Algorithms/234_Palindrome_Linked_List/Palindrome_Linked_List.java
import java.util.ArrayList;

/*
描述：回文链表
请判断一个链表是否为回文链表。

示例 1:

输入: 1->2
输出: false
示例 2:

输入: 1->2->2->1
输出: true
进阶：
你能否用 O(n) 时间复杂度和 O(1) 空间复杂度解决此题？

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/palindrome-linked-list
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
*/

/**
 * Definition for singly-linked list.
 */
 public class ListNode {
	 int val;
	 ListNode next;
     ListNode(int x) { val = x; }
}
 
class Solution {
    public boolean isPalindrome(ListNode head) {
        //
		if(head == null || head.next == null){
			return true;
	}
    	ArrayList<Integer> res = new ArrayList<Integer>();
    	while(head != null)
    		res.add(head.val);
    		head = head.next;
		int i = 0;
		int j = arr.size()-1;
    	while(i<j) {
			if(arr.get(i).compareTo(arr.get(j))!=0) {
				return false;
			}
			++i;
			--j;
		}
		return true;
    }
}